### PR TITLE
Submission Record Series: BatchOpt+MLP4+RoPE100k and 8L EMA Int6 Bigram65k on Single 20GB GPU (val_bpb 1.7810 → 1.3092)

### PR DESCRIPTION
**Pull request by markste-in**
This MR bundles two consecutive improvements for the 10min/16MB track, all developed and validated on a single small GPU (20GB VRAM) under the 600s wall-clock constraint.

1) 2026-03-22 — BatchOpt + MLP4 + RoPE100k

Improved val_bpb from 1.7810 → 1.4784 (+16.9%).
Main drivers: batch reduction (196k → 98k tokens) to increase update count, MLP_MULT=4, LR retune (0.035), shorter warmdown (600), and ROPE_BASE=100k.
Result: strong quality gain while staying comfortably within size/runtime limits.
2) 2026-03-23 — 8L + EMA + Int6-MLP + BigramHash-65k

Further improved to 1.3092 (1.3978 → 1.3092 in session-2, +6.3%).
Main drivers: 8-layer architecture, BigramHash vocab scaled to 65k, EMA replacing SWA, longer linear warmdown (3000), stride-64 sliding eval, int6 MLP quantization, and 1% magnitude pruning.
Final artifact: 15.93MB (within 16MB cap).
Net effect: a staged progression from baseline 1.7810 to final 1.3092, achieved with practical single-GPU (20GB) optimization rather than large-cluster training.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Submission Record Series: BatchOpt+MLP4+RoPE100k and 8L EMA Int6 Bigram65k on Single 20GB GPU (val_bpb 1.7810 → 1.3092)

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache


## PR #759 Analysis

**Files reviewed:**
- `records/track_10min_16mb/2026-03-22_BatchOpt_MLP4_RoPE100k/train_gpt.py` (1130 lines)
- `records/track_10min_16mb/2026-03-23_8L_EMA_Int6_Bigram65k/train_gpt.py` (1350 lines)

---

### 1. N-gram family bug — NOT FOUND

The newer submission (8L_EMA_Int6_Bigram65k) includes a `BigramHashEmbedding` class. The hash function at line 641:

```python
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```

This XORs adjacent **input** tokens (`t[i]` XOR `t[i-1]`) — the legal BigramHash pattern. The target token is not included in the hash key. No `primes[k]`, no `target * primes[k]`, no `ctx_hash ^ target`. CLEAN.

The first submission (BatchOpt_MLP4_RoPE100k) has no hash structures at all.

---

### 2. Pre-Quant TTT on val_tokens — NOT FOUND

Neither file contains any TTT (test-time training) logic. There is no `optimizer.step()` call during evaluation, no multi-epoch AdamW on val_tokens. The validation functions (`eval_val` at line 229/255 and `eval_val_sliding` at line 310) both use `torch.no_grad()` / `torch.inference_mode()` exclusively.

---

### 3. Legal score-first-per-chunk TTT — NOT FOUND (not applicable)

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
